File: baitaptuan3/C++/quanlyungdung.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

class TextEditor {
private:
    string text;
    stack<string> undo_stack;
    stack<string> redo_stack;

public:
    void type_text(const string& new_text) {
        undo_stack.push(text);
        text += new_text;
        while (!redo_stack.empty()) redo_stack.pop();  // Clear redo stack
        cout << "Da go: " << new_text << endl;
        display();
    }

    void undo() {
        if (!undo_stack.empty()) {
            redo_stack.push(text);
            text = undo_stack.top();
            undo_stack.pop();
            cout << "Da hoan tac hanh dong." << endl;
            display();
        } else {
            cout << "Khong co hanh dong nao de hoan tac." << endl;
        }
    }

    void redo() {
        if (!redo_stack.empty()) {
            undo_stack.push(text);
            text = redo_stack.top();
            redo_stack.pop();
            cout << "Da lam lai hanh dong." << endl;
            display();
        } else {
            cout << "Khong co hanh dong nao de lam lai." << endl;
        }
    }

    void display() const {
        cout << "Noi dung hien tai: " << text << endl;
    }
};

// Minh họa
int main() {
    TextEditor editor;
    editor.type_text("Hello");
    editor.type_text(" World");
    editor.undo();
    editor.undo();
    editor.redo();
    return 0;
}

File: Buoi 3/Julia/quanlyungdung.jl
mutable struct TextEditor
    text::String
    undo_stack::Array{String,1}
    redo_stack::Array{String,1}
end

function TextEditor()
    return TextEditor("", String[], String[])
end

function type_text(editor::TextEditor, new_text::String)
    push!(editor.undo_stack, editor.text)
    editor.text *= new_text
    empty!(editor.redo_stack)
    println("Da go: $new_text")
    display(editor)
end

function undo(editor::TextEditor)
    if !isempty(editor.undo_stack)
        push!(editor.redo_stack, editor.text)
        editor.text = pop!(editor.undo_stack)
        println("Da hoan tac hanh dong.")
        display(editor)
    else
        println("Khong co hanh dong nao de hoan tac.")
    end
end

function redo(editor::TextEditor)
    if !isempty(editor.redo_stack)
        push!(editor.undo_stack, editor.text)
        editor.text = pop!(editor.redo_stack)
        println("Da lam lai hanh dong.")
        display(editor)
    else
        println("Khong co hanh dong nao de lam lai.")
    end
end

function display(editor::TextEditor)
    println("Noi dung hien tai: $(editor.text)")
end

# Minh họa
editor = TextEditor()
type_text(editor, "Hello")
type_text(editor, " World")
undo(editor)
undo(editor)
redo(editor)
